### ISSUE TITLE
RAM overload with more than 600K results

### ISSUE THREAD

**Issue by SqrtMinusOne** [liuchengxu/vim-clap#292]
**Environment (please complete the following information):**
 - OS: Linux Mint 19.2
 - NeoVim version: 0.4.3
 - vim-clap version: commit f459551
 - Have you reproduced with a minimal vimrc: No
 - Have you updated to the latest plugin version: Yes
 - Have you upgraded to/compiled the latest Rust binary: Yes

**Describe the bug**
I frequently use the global fuzzy-finder. Sometimes ripgrep outputs an unholy amount of results, e.g. in one of the projects I'm currently working on an output of
`rg -H --no-heading --vimgrep --smart-case "$cookie" | wc -l`
would be `619660` (619660 lines in the result). I wanted to find places where I use `$cookie` and forgot that `$` is a special symbol in regex, so certain lines were repeated multiple times.

Ripgrep can handle this, but vim-clap consumes a lot of RAM while receiving such a large input. 200K lines is fine, but 600K lines is more than my computer can handle.

In essence, if I type something wrong in `Clap grep`, I risk crashing my PC.

**Clap debug**
```
has cargo:             1
has maple:             /home/pavel/.local/share/nvim/plugged/vim-clap/target/release/maple
maple info:            version 0.1.7, built for x86_64-unknown-linux-gnu by rustc 1.40.0 (73528e339 2019-12-16).
has +python3:          1
has py dynamic module: 0
Current FileType:      startify
Third Party Providers: []

Global Options:
let g:clap#autoload_dir              = '/home/pavel/.local/share/nvim/plugged/vim-clap/autoload'
let g:clap#provider_alias            = {'gfiles': 'git_files', 'hist:': 'command_history'}
let g:clap_disable_bottom_top        = 0
let g:clap_enable_icon               = 1
let g:clap_forerunner_status_sign    = {'done': '•', 'running': '!'}
let g:clap_layout                    = {'relative': 'editor'}
let g:clap_no_matches_msg            = 'NO MATCHES FOUND'
let g:clap_open_action               = {'ctrl-v': 'vsplit', 'ctrl-x': 'split', 'ctrl-t': 'tab split'}
let g:clap_search_box_border_style   = 'nul'
let g:clap_search_box_border_symbols = {'nil': ['', ''], 'curve': ['', ''], 'arrow': ['', '']}

Provider Variables:
[]
```
**To Reproduce**
Run `Clap grep` with some crappy regex like `$cookie` in a moderately large project. Or `Clap` with any provider which returns more than 600K results.

**Expected behavior**
I'd expect the memory not to be overloaded.

The easiest solution I can see is to make an option to pipe results to `| head -n {number of lines}` before processing them in the plugin. That way the number of results would be limited.

**Comment by liuchengxu** [comment]
I'm guessing you're using a computer with small memory as 600K is actually not huge. My computer has 32GB so I seldom ran into this issue, but the problem does exist, for the whole output of command is keep in memory right now https://github.com/liuchengxu/vim-clap/blob/d665610884/src/main.rs#L132 . I know it's rough but works for me, I'm not sure when this issue will be resolved.

Now you can try using the job of neovim directly, see if that helps.

```diff
diff --git a/autoload/clap/provider/grep.vim b/autoload/clap/provider/grep.vim
index 17a4516..bdff24f 100644
--- a/autoload/clap/provider/grep.vim
+++ b/autoload/clap/provider/grep.vim
@@ -132,7 +132,8 @@ function! s:spawn(query) abort
 
   call clap#rooter#try_set_cwd()
 
-  if clap#maple#is_available()
+  " if clap#maple#is_available()
+  if 0
     let [grep_opts, query] = s:translate_query_and_opts(a:query)
     " Add ' .' for windows in maple
     call clap#maple#run_grep(s:grep_executable.' '.grep_opts, query, s:grep_enable_icon, s:ripgrep_glob)
```

Or you could use the other more mature tools, e.g., fzf.vim.

**Comment by SqrtMinusOne** [comment]
That did help with 600K results.

However, I switched to another branch, and there even running riprgrep from the command line causes an overflow. Vim-clap said the number of results exceeded 1M before I pressed Escape. I think I'll raise an issue to ripgrep after I localize such a behavior. It's unlikely that any modification of the plugin can help with this.

I guess I'll be more careful with what I type, and maybe switch to ag if I often get something like this.
